server/lib/tuist/qa/agent.ex

```diff
@@ -3,33 +3,212 @@ defmodule Tuist.QA.Agent do
   Tuist QA agent module.
   """
 
+  alias LangChain.Chains.LLMChain
+  alias LangChain.ChatModels.ChatAnthropic
+  alias LangChain.Message
+  alias LangChain.Message.ContentPart
+  alias LangChain.Message.ToolResult
+  alias LangChain.TokenUsage
+  alias Tuist.QA.Tools
   alias Tuist.Simulators
   alias Tuist.Zip
 
-  def test(%{preview_url: preview_url, bundle_identifier: bundle_identifier}) do
-    run_preview(preview_url, bundle_identifier)
+  require Logger
+
+  def test(%{preview_url: preview_url, bundle_identifier: bundle_identifier, prompt: prompt}) do
+    with {:ok, simulator_device} <- simulator_device(),
+         :ok <- run_preview(preview_url, bundle_identifier, simulator_device) do
+      handler = %{
+        on_message_processed: fn _chain, %Message{content: content} ->
+          for %ContentPart{type: :text, content: content} <- content || [] do
+            Logger.debug(content)
+          end
+        end,
+        on_llm_token_usage: fn _chain, %TokenUsage{input: input, output: output} ->
+          Logger.debug("LLM token usage: #{input} input tokens, #{output} output tokens")
+        end
```

Comment:
these are really useful as we develop the agent. Both should be eventually sent as events as part of the agent telemetry. The token usage can be used to couple the costs with given customers in case they'd be using our API token.